register-machines/src/main.rs

```diff
@@ -8,44 +8,114 @@ type State = (Label, HashMap<Register, u128>);
 enum Instruction {
     Add(Register, Label),
     Sub(Register, Label, Label),
-    Halt
+    Halt,
 }
-
 use Instruction::*;
+
+#[allow(dead_code)]
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut label, mut regs) = state.clone();
+    while (0..program.len()).contains(&label) {
+        match program[label] {
+            Add(reg, new_label) => {
+                *regs.entry(reg).or_default() += 1;
+                label = new_label;
+            }
+            Sub(reg, new_label, other_label) => {
+                let register = regs.entry(reg).or_default();
+                label = if *register > 0 {
+                    *register -= 1;
+                    new_label
+                } else {
+                    other_label
+                }
+            }
+            Halt => break,
+        };
+    }
+    (label, regs)
 }
 // <<x,y>> = (2^x)*(2y+1)
+#[allow(dead_code)]
 fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+    (1 << x) * (2 * y + 1)
 }
 // <x,y> = (2^x)*(2y+1)-1
+#[allow(dead_code)]
 fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+    encode_pair1(x, y) - 1
 }
+#[allow(dead_code)]
 fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+    match l {
+        [] => 0,
+        [x, l @ ..] => encode_pair1(*x, encode_list_to_godel(l)),
+    }
 }
```

Comment:
This code can't really be optimised by the compiler:
```
example::encode_list_to_godel:
        test    rsi, rsi
        je      .LBB1_1
        push    r14
        push    rbx
        push    rax
        lea     rax, [rdi + 16]
        add     rsi, -1
        mov     rbx, qword ptr [rdi]
        mov     r14, qword ptr [rdi + 8]
        mov     rdi, rax
        call    qword ptr [rip + example::encode_list_to_godel@GOTPCREL]
        mov     rcx, rdx
        mov     rdi, rbx
        mov     rsi, r14
        mov     rdx, rax
        add     rsp, 8
        pop     rbx
        pop     r14
        jmp     example::encode_pair1
.LBB1_1:
        xor     eax, eax
        xor     edx, edx
        ret
```

However, this might noticeably be a case for a fold with an iterator:
```
l.iter().rev().fold(0, |acc, v| encode_pair1(*v, acc))
```
which produces:
```
example::encode_list_to_godel:
        push    r14
        push    rbx
        push    rax
        test    rsi, rsi
        je      .LBB1_1
        mov     rbx, rsi
        mov     r14, rdi
        shl     rbx, 4
        xor     eax, eax
        xor     ecx, ecx
.LBB1_3:
        mov     rdi, qword ptr [r14 + rbx - 16]
        mov     rsi, qword ptr [r14 + rbx - 8]
        mov     rdx, rax
        call    example::encode_pair1
        mov     rcx, rdx
        add     rbx, -16
        jne     .LBB1_3
        jmp     .LBB1_4
.LBB1_1:
        xor     eax, eax
        xor     ecx, ecx
.LBB1_4:
        mov     rdx, rcx
        add     rsp, 8
        pop     rbx
        pop     r14
        ret
```
It's kinda nuts that this also optimises away any bounds checking if you can tell. I'm honestly super impressed with the power of iterators in this language. Anyways: the lesson of the day is to think iterators and higher-order functions before recursion in rust.